File: Prism/Old Javascript/Timers.js
var Timers = function () {
    var clock = 0.0;
    var limit = PrismSheet.getRange("B3").getValue();
    var activeCooldowns = [];

    function cyclone() {
        var cycleArray = [];
        cycleArray.push(activeCooldowns);
        var allies = Targets.getAllies();
        var enemies = Targets.getEnemies();
        for (var ally in allies) {
            cycleArray.push(allies[ally].Effects);
        }
        for (var enemy in enemies) {
            cycleArray.push(enemies[enemy].Effects);
        }
        return cycleArray;
    };

    function cycle(time) {
        var cycleArray = cyclone();
        for (var c = 0; c < cycleArray.length; c++) {
            var innerArray = cycleArray[c];
            if (innerArray.length > 0) {
                for (var i = innerArray.length - 1; i >= 0; i--) {
                    if (innerArray[i][1] > 0) {
                        innerArray[i][1] -= time;
                    }
                    if (innerArray[i][1] <= 0) {
                        innerArray.splice(i, 1);
                    }
                }
            }
        }
    };
    return {
        //Used to keep the main loop operating
        isRunning: function () {
            return clock < limit;
        },
        cycleTimers: function (time) {
            clock += time;
            cycle(time);
        },
        getClock: function () {
            return clock;
        },
        //Takes spell string name and number cooldown
        initiateCooldown: function (name, cooldown) {
            if (cooldown > 0) {
                activeCooldowns.push([name, cooldown]);
            }
        },
        //Takes spell string name
        isSpellOnCooldown: function (name) {
            if (activeCooldowns.length > 0) {
                for (var i = 0; i < activeCooldowns.length; i++) {
                    if (activeCooldowns[i][0] == name) {
                        return true;
                    }
                }
            }
            return false;
        },
        resetCooldown: function (name) {
            if (activeCooldowns.length > 0) {
                for (var i = 0; i < activeCooldowns.length; i++) {
                    if (activeCooldowns[i][0] == name) {
                        activeCooldowns.splice(i, 1);
                    }
                }
            }
        }
    };
}();

File: Prism/Basic Classes/Timers.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Prism.Basic_Classes
{
    public class Timers
    {   
        private List<Cooldown> _activeCooldowns;

        private decimal _clock = 0;
        public decimal Clock
        {
            get
            {
                return _clock;
            }
        }

        private decimal _limit = 180;
        public decimal Limit
        {
            get
            {
                return _limit;
            }
        }

        public void CycleActiveCooldowns(decimal time)
        {
            Cycle(ref _activeCooldowns, time);
        }

        public static void Cycle(ref List<Cooldown> timers, decimal time)
        {
            foreach (Cooldown timer in timers)
            {
                timer.CooldownTime -= time;
                if (timer.CooldownTime <= 0) timers.Remove(timer);
            }
        }

        public bool IsRunning()
        {
            return Clock < Limit;
        }

        public void CycleTimers(decimal time)
        {
            _clock += time;
            CycleActiveCooldowns(time);
        }

        public void InitiateCooldown(string name, decimal cooldown)
        {
            _activeCooldowns.Add(new Cooldown(name, cooldown));
        }

        public bool IsCooldown(string name)
        {
            return _activeCooldowns.Exists(x => x.Name == name);
        }

        public void ResetCooldown(string name)
        {
            _activeCooldowns.Remove(_activeCooldowns.Find(x => x.Name == name));
        }
    }
}
